src/osd/OSDMap.cc

```diff
@@ -6226,39 +6517,181 @@ int OSDMap::calc_read_balance_score(CephContext *cct, int64_t pool_id,
-                           prim_affinity_sum, num_osds, osd_pa_count,
-                           total_osd_weight, max_prims_per_osd,
-                           max_acting_prims_per_osd, avg_prims_per_osd,
-                           prim_on_zero_pa, acting_on_zero_pa, max_osd_score);
+  int rc = tmp_osd_map.set_rbi_fair(cct, rbi, pool_id, total_weighted_pa,
+                                    prim_affinity_sum, num_osds, osd_pa_count,
+                                    total_osd_weight, max_prims_per_osd,
+                                    max_acting_prims_per_osd, avg_prims_per_osd,
+                                    prim_on_zero_pa, acting_on_zero_pa, max_osd_score);
 
   if (cct != nullptr) {
     ldout(cct,30) << __func__ << " pool " << get_pool_name(pool_id)
-                  << " pa_avg " << p_rbi->pa_avg
-                  << " pa_weighted " << p_rbi->pa_weighted
-                  << " pa_weighted_avg " << p_rbi->pa_weighted_avg
-                  << " optimal_score " << p_rbi->optimal_score
-                  << " adjusted_score " << p_rbi->adjusted_score
-                  << " acting_adj_score " << p_rbi->acting_adj_score
+                  << " pa_avg " << rbi.pa_avg
+                  << " pa_weighted " << rbi.pa_weighted
+                  << " pa_weighted_avg " << rbi.pa_weighted_avg
+                  << " optimal_score " << rbi.optimal_score
+                  << " adjusted_score " << rbi.adjusted_score
+                  << " acting_adj_score " << rbi.acting_adj_score
                   << dendl;
     ldout(cct,20) << __func__ << " pool " << get_pool_name(pool_id)
-		  << " raw_score: " << p_rbi->raw_score
-		  << " acting_raw_score: " << p_rbi->acting_raw_score
+		  << " raw_score: " << rbi.raw_score
+		  << " acting_raw_score: " << rbi.acting_raw_score
 		  << dendl;
     ldout(cct,10) << __func__ << " pool " << get_pool_name(pool_id)
-		  << " wl_score: " << p_rbi->acting_adj_score << dendl;
+		  << " wl_score: " << rbi.acting_adj_score << dendl;
   }
 
   return rc;
+
+}
+
+int OSDMap::calc_rbs_size_optimal(CephContext *cct, OSDMap& tmp_osd_map, int64_t pool_id,
+                                  const pg_pool_t *pgpool, read_balance_info_t &rbi) const
+{
+  zero_rbi(rbi);
+  rbi.score_type = RBS_SIZE_OPTIMAL;
+
+  auto num_pgs = pgpool->get_pg_num();
+
+  if (num_pgs == 0) {
+    rbi.err_msg = fmt::format("ERROR: pool {} has no PGs", pool_id);
+    if (cct != nullptr) {
+      ldout(cct, 20) << __func__ << " pool " << pool_id
+                     << " has no PGs - can't calculate size-optimal read balancer score" << dendl;
+    }
+  }
+
+  map<uint64_t,set<pg_t>> pgs_by_osd;
+  map<uint64_t,set<pg_t>> prim_pgs_by_osd;
+  map<uint64_t,set<pg_t>> acting_prims_by_osd;
+
+  pgs_by_osd = tmp_osd_map.get_pgs_by_osd(cct, pool_id, &prim_pgs_by_osd, &acting_prims_by_osd);
+  auto num_osds = pgs_by_osd.size();
+  int64_t num_pg_osd_legs = 0;
+  for (int i = 0 ; i < num_osds ; i++) {
```

Comment:
Also seeing a compile warning on this line. `num_osds` should probably be cast to an `int`:
```
/home/lflores/ceph/src/osd/OSDMap.cc:6579:22: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
 6579 |   for (int i = 0 ; i < num_osds ; i++) {
      |                    ~~^~~~~~~~~~
```